### PR TITLE
feat: snapshot status indicator

### PR DESCRIPTION
**Pull request by 1emu**
Closes [#1249](https://github.com/decentraland/governance/issues/1249)

image

image

Mobile view with menu open

image

image

image

### CODE REVIEW

File: src/services/SnapshotService.ts [hunk 1]
```diff
@@ -3,20 +3,62 @@ import isNumber from 'lodash/isNumber'
...
+      let health = ServiceHealth.Normal
+      const responseTime = await SnapshotGraphql.get().ping()
+      if (responseTime === -1) {
+        health = ServiceHealth.Failing
+      } else if (responseTime > SLOW_RESPONSE_TIME_THRESHOLD_IN_MS) {
+        health = ServiceHealth.Slow
+      }
+
+      const snapshotStatus = { health, responseTime }
+      this.saveStatusInCache(snapshotStatus)
```
Comment (andyesp) [inline src/services/SnapshotService.ts:48]: If we are not reusing this method I'd call CacheService directly here as the method is a one-liner.

File: src/services/SnapshotService.ts [hunk 2]
```diff
@@ -3,20 +3,62 @@ import isNumber from 'lodash/isNumber'
...
+import { ErrorCategory } from '../utils/errorCategories'
 
+import CacheService from './CacheService'
+import { ErrorService } from './ErrorService'
 import { ProposalInCreation, ProposalLifespan } from './ProposalService'
 import RpcService from './RpcService'
 
 const DELEGATION_STRATEGY_NAME = 'delegation'
+const SLOW_RESPONSE_TIME_THRESHOLD_IN_MS = 5000 // 5 seconds
+const PING_INTERVAL_IN_SECONDS = 60 // 1 minute
```
Comment (andyesp) [inline src/services/SnapshotService.ts:29]: Maybe `STATUS_CACHE_TIMEOUT_IN_SECONDS`?

File: src/clients/SnapshotApi.ts [hunk 3]
```diff
@@ -188,4 +185,27 @@ export class SnapshotApi {
...
+      const startTime = now.getTime()
+      await snapshot.utils.getScores(spaceName, strategies, network, formattedAddresses, undefined, scoreApiUrl)
+      const endTime = new Date().getTime()
+      return endTime - startTime
+    } catch (error) {
+      return -1 // Return -1 to indicate API failures
+    }
+  }
+
+  private async prepareArgs(addresses: string[]) {
```
Comment (andyesp) [inline src/clients/SnapshotApi.ts:203]: Maybe `prepareScoresQueryArgs`? or `prepareScoresUtilsArgs`

File: src/clients/SnapshotGraphql.ts [hunk 4]
```diff
@@ -444,23 +443,31 @@ export class SnapshotGraphql extends API {
...
+        GRAPHQL_ENDPOINT,
+        this.options()
+          .method('POST')
+          .json({
+            query: GET_VOTES_QUERY,
+            variables: {
+              space: SNAPSHOT_SPACE,
+              first: 10,
+              start: getQueryTimestamp(getAMonthAgo(now).getTime()),
+              end: getQueryTimestamp(now.getTime()),
```
Comment (andyesp) [inline src/clients/SnapshotGraphql.ts:460]: `getQueryTimestamp(startTime)`

File: src/components/Debug/SnapshotStatus.tsx [hunk 5]
```diff
@@ -1,55 +1,51 @@
...
-          </Button>
-        </div>
-      </ContentSection>
-      <Label>{'Status'}</Label>
-      <Text>{JSON.stringify(snapshotStatus)}</Text>
-      <Label>{'Space'}</Label>
-      <Text>{JSON.stringify(snapshotSpace)}</Text>
-      {!!errorMessage && <ErrorMessage label={'Snapshot Error'} errorMessage={errorMessage} />}
-    </div>
+    <>
```
Comment (andyesp) [inline src/components/Debug/SnapshotStatus.tsx:42]: Fragment not necessary

File: src/services/SnapshotService.ts [hunk 6]
```diff
@@ -3,20 +3,79 @@ import isNumber from 'lodash/isNumber'
...
+
+      logger.log('Snapshot status:', snapshotStatus)
+      return snapshotStatus
+    } catch (error) {
+      ErrorService.report('Unable to determine snapshot status', { error, category: ErrorCategory.Snapshot })
+      return undefined
+    }
+  }
+
+  private static async sense(): Promise<SnapshotStatus> {
```
Comment (andyesp) [inline src/services/SnapshotService.ts:51]: I think we could go with the "ping" nomenclature as this is used internally in this method by the snapshot clients and we could infer a ping is a status check.

**Review by andyesp — APPROVED** [review]

**Review by andyesp — APPROVED** [review]

**Review by ncomerci — APPROVED** [review]

**Pull request merged**